test/process/io/SourceTest.cpp

```diff
@@ -0,0 +1,124 @@
+const char *TEST_STREAM_PSOURCE_OPERATIONS = "Test PathSource operations";
+
+path getAndCheckTextFile() {
+    path textFile = TestResourcesAndFunctions::getResource("TestTextFile.txt");
+            CHECK(exists(textFile));
+            CHECK(file_size(textFile) == 67);
+    return textFile;
+}
+
+void runInputStreamTest(Source *Source, int64_t expectedSize) {
+    /**
+     * File with the following content:
+     * First line
+     * Second line
+     * Third line
+     * Fourth line
+     * Fifth line
+     * Sixth line
+     */
+    Byte buf[128]{0};
+    Source->open();
+    auto readBytes = Source->read(buf, 5);
+            CHECK(readBytes == 5);
+            CHECK(Source->size() == expectedSize);
+            CHECK(string(reinterpret_cast<const char *>( buf)) == string("First"));
+    memset(buf, 0, 128);
+            CHECK(Source->skip(1) != 0);
+            CHECK(Source->readChar() == 'l'); // l of line
+            CHECK(Source->skip(4) != 0);
+            CHECK(Source->readChar() == 'S'); // S of Second
+            CHECK(Source->tell() == 12);
+    Source->seek(0, true);
+            CHECK(Source->tell() == 0);
+            CHECK(Source->readChar() == 'F'); // F of First
+            CHECK(Source->close());
+}
+
+SUITE (SUITE_BIS_TESTS) {
+    TEST (TEST_STREAM_PSOURCE_OPERATIONS) {
+        TestResourcesAndFunctions res(SUITE_BIS_TESTS, TEST_STREAM_PSOURCE_OPERATIONS);
+        path textFile = getAndCheckTextFile();
+
+        PathSource Source(textFile);
+                CHECK(Source.isFile());
+
+
+        runInputStreamTest(&Source, file_size(textFile));
+    }
+
+    TEST (testCanRead) {
+        TestResourcesAndFunctions res(SUITE_BIS_TESTS, TEST_STREAM_ISOURCE_OPERATIONS);
+        path testFile = getAndCheckTextFile();
+        ifstream testData(testFile);
+
+        StreamSource Source(&testData);
+
+                CHECK(Source.tell() == 0);
+                CHECK(Source.canRead());
+                CHECK(Source.tell() == 0);
+    }
+
+    TEST (testCanReadWontWorkWhenDatasourceIsDepleted) {
+        TestResourcesAndFunctions res(SUITE_BIS_TESTS, TEST_STREAM_ISOURCE_OPERATIONS);
+        path testFile = getAndCheckTextFile();
+        ifstream testData(testFile);
+
+        Bytef buf[32768]{0};
+        StreamSource Source(&testData, 4, 4);
+        auto read = Source.read(buf, 32768);
+                CHECK(!Source.canRead());
+                CHECK(read == static_cast<int64_t>(file_size(testFile)));
+
+        // Also, canRead can also be true, if a valid rewind was done.
+//                CHECK(false);
+    }
+
+    TEST (TEST_STREAM_ISOURCE_OPERATIONS) {
+        TestResourcesAndFunctions res(SUITE_BIS_TESTS, TEST_STREAM_ISOURCE_OPERATIONS);
+        ifstream testData(getAndCheckTextFile());
+
+        StreamSource Source(&testData);
+                CHECK(Source.isStream());
+
+        int64_t expectedSize = -1; // Use overflow to get the max value.
```

Comment:
There should be a constant in the C++ standard-lib. I am not sure whether this overflow approach is portable.